### ISSUE TITLE
Retry SubmitForm526 Jobs When They Have Multiple BIRLS IDs

### ISSUE THREAD

**Issue by edmangimelli** [department-of-veterans-affairs/vets-api#5268]
# tl;dr

Sometimes _SubmitForm526_ jobs fail because MPI (formerly MVI) has multiple active BIRLS IDs for the Veteran (we always just grab the first BIRLS ID ([IdParser](https://github.com/department-of-veterans-affairs/vets-api/blob/ae1a0da535fe7427b3936bf47c44c6de0fe53fff/lib/mvi/responses/id_parser.rb)))

I propose that, upon a jobs failure (exhaustion or non-retryable error), we check to see if the Veteran has multiple BIRLS, and, if they do, retry the job with a different BIRLS ID.

Flow:

```
if a job fails (exhausts or non-retryable error)
   get ALL birls ids for veteran
   have we tried all of these?
   yes
      DONE
   no
      put current birls id into birls_ids_tried (new encrypted DB column (array))
      update auth_headers with new birls
      start new submit job (with new birls)
```

**`This will require changes to our error reporting.`** See "Changes to Error Reporting" below.

# The Issue

Some error messages received from EVSS during Form526 submission can be fairly opaque. While investigating "external service unavailable" errors, I discovered that many of these fail because a system downstream from EVSS is rejecting submissions when the Veteran has bad data in MPI. "external service unavailable" errors specifically seem to be caused by a Veteran having multiple active BIRLS IDs (these seem to often be typos with a couple characters transposed or one character different).

Using the following snippet, I tested a variety of errors ("external service unavailable" errors and a variety of cryptic 500s from EVSS), and discovered that the multiple BIRLS problem might be causing many of our submit jobs to fail with a 500. 

Snippet:
```
def job_id_to_active_birls_ids(job_id)
  job_statuses = Form526JobStatus.where job_id: job_id
    raise "couldn't find job" if job_statuses.count == 0
    raise "found multiple jobs" if job_statuses.count > 1
  submission_to_active_birls_ids job_statuses.first.submission
end

def submission_to_active_birls_ids submission
  submission = Form526Submission.find(submission) if submission.is_a?(String) || submission.is_a?(Integer)
  auth_headers = submission.auth_headers
  edipi = auth_headers['va_eauth_dodedipnid']
    raise "edipi nil" if edipi.nil?
  accounts = Account.where edipi: edipi
    raise "edipi pulled up multiple accounts" if accounts.count > 1
  icn = accounts.first&.icn
  service = MVI::Service.new
  user_identity =  OpenStruct.new mhv_icn: icn, dslogon_edipi: edipi
  profile_message = service.send :create_profile_message, user_identity
  mvi_response = service.send :perform, :post, '', profile_message, soapaction: 'PRPA_IN201305UV02'
  body_xpath = 'env:Envelope/env:Body/idm:PRPA_IN201306UV02'
  body = mvi_response.body.locate(body_xpath).first
  subject_xpath = 'controlActProcess/subject'
  subject = body.locate(subject_xpath).first
  raise unless subject
  patient_xpath = 'registrationEvent/subject1/patient'
  patient = subject.locate(patient_xpath).first
  raise unless patient
  va_root_oid = '2.16.840.1.113883.4.349'
  MVI::Responses::IdParser.new.select_ids_with_extension(
    patient.locate('id'), /^\w+\^PI\^200BRLS\^USVBA\^A$/,
    va_root_oid
  )
end

def compare_strings(strings)
  longest_string = strings.max_by(&:length)
  
  comparison = ''
  longest_string.length.times do |i|
    comparison << (strings.map { |string| string[i] }.uniq.length == 1 ? ' ' : 'x')
  end

  puts
  puts comparison.blank? ? 'no differences' : 'different!'
  puts
  puts [comparison, *strings]
  puts
  puts
end

compare_strings job_id_to_active_birls_ids 'some_job_id_here'
```

# Proposed Solution

I propose that, upon a job's failure (exhaustion or non-retryable error), we look for multiple BIRLS, and, if there are any, retry the job with the other BIRLS ID. I propose we add an encrypted column to Form526Submissions that will keep track of the BIRLS IDs we've tried.

# PRs for this:

[Main one](https://github.com/department-of-veterans-affairs/vets-api/pull/5245)

[adding the DB columns](https://github.com/department-of-veterans-affairs/vets-api/pull/5234)

[retrieving all BIRLS](https://github.com/department-of-veterans-affairs/vets-api/pull/5228)

# Notes on the Implementation

## Went through several iterations on this

Originally wanted to leave the Form526Submission database record untouched and send the alternate BIRLS ID as a parameter to the sidekiq job's `perform` method. Sidekiq uses the parameters you pass in as an identifier for the job, and, consequently, those parameters (JSONified) get stored in redis. Because I don't want to store a BIRLS ID unencrypted, I opted to not go this route, which is why I decided to keep track of with BIRLS IDs had been tried on the Form526Submission record itself.

In my second iteration, I was going to have retrieval of BIRLS IDs happen only once (after the first failure), and all subsequent attempts would refer to that list (however stored in the DB). I opted to not got this route in order to keep things as simple as possible. Now, any failure does a BIRLS ID lookup, checks to see if there are any BIRLS IDs that haven't been tried, and proceeds accordingly. It's more lookups but a far simpler and easier to understand flow. This is the same reason why I didn't simply put a retry in the _non-retryable error handler_ (if there were more BIRLS) --I was going to need to start a new job if there were multiple BIRLS and the original job exhausted, I therefore thought it would be simpler to start a new job if the original job had hit a non-retryable error --that is, all failures kick off the same flow.

# Changes to Error Reporting

The biggest drawback to my proposal is that determining what submissions are successful / failed will no longer be as straightforward.

Right now, a submission has ***1*** submit job, with sidekiq completely handling the retries, and 1 job status record which is continually updated until the submit job is successful or failed. 

My proposal would cause some submissions to have ***multiple*** submit jobs (although **never** running at the same time), with each having their own job status record.

Therefore, to determine whether or not a submission was successful, you would have to look at multiple submit-job statuses --you can't just look at the first one returned.

**This means that my solution will not cause the number of failures to go down, but it** ***will*** **cause the number of successes to go up.**

Consequently, we'll have to adjust our error reporting tools (rake tasks, etc), to take this into account.

**Comment by annaswims** [comment]
@edmangimelli is this the error we see when we have a bad birls id? https://github.com/department-of-veterans-affairs/va.gov-team/issues/11353
